Lib/pathlib.py

```diff
@@ -347,7 +126,8 @@ def _select_from(self, parent_path, is_dir, exists, scandir):
 class _WildcardSelector(_Selector):
 
     def __init__(self, pat, child_parts, flavour):
-        self.match = flavour.compile_pattern(pat)
+        flags = re.NOFLAG if flavour is posixpath else re.IGNORECASE
+        self.match = re.compile(fnmatch.translate(pat), flags).fullmatch
```

Comment:
Isn’t case sensitivity a property of each filesystem, not the OS?

Reply:
🤔 yes, but Python (including `os.path` and `pathlib`) makes a hard division between Windows (where we normalise case before making comparisons etc) and POSIX (inc. Mac). Here I've preserved existing behaviour. But @eryksun has already found at least one instance where pathlib gets it wrong: #86824

Reply:
@barneygale, NTFS in Windows 10 allows enabling case sensitivity for directory trees, so it's no longer strictly a property of either the filesystem or the OS. Using `ntpath.normcase()`, among other case-insensitive comparisons, is a strategy that's increasingly running into problems. For some cases, such as wildcard matching in a directory listing (when not using `FindFirstFileW()` for a DOS-style wildcard match), one should query whether the listed directory is case sensitive via `GetFileInformationByHandleEx()`: `FileCaseSensitiveInfo` (23). In other cases, such as remembering seen symlink targets, the true normalized path is needed. Call `CreateFileW()` to open the path with reparsing disabled (i.e. `FILE_FLAG_OPEN_REPARSE_POINT`), and then call `GetFinalPathNameByHandleW()` to get the final normalized path.

Reply:
Thank you Eryk. I'd like to log a follow-up bug so we don't lose track of this info, but it's a bit beyond my current level of Windows expertise. It feels like too large a job to solve here. @merwok how should we proceed do you think?

Reply:
Nowadays, Linux EXT4 can be formatted to enable [per-directory case folding](https://www.collabora.com/news-and-blog/blog/2020/08/27/using-the-linux-kernel-case-insensitive-feature-in-ext4/), similar to how this works with NTFS in Windows. The assumption about case sensitivity in POSIX and Windows is wide spread throughout the standard library and import system (e.g. `sys.path`). Supporting per-directory case sensitivity would be a lot of work. I don't know whether it's worth it, but I just wanted to point out the increasing set of cases where our assumption breaks down.

Reply:
I see!  OK to keep applying the current division then.  Could someone check if the docs are clear about this assumption?

Reply:
*Fairly* clear IMO. In `os.path`:

![image](https://user-images.githubusercontent.com/960340/166457363-4f98d891-ffa2-497a-ba43-7cf62440ef17.png)

![image](https://user-images.githubusercontent.com/960340/166457427-edd6d0ca-36fa-4902-96e8-5419c69e3a8a.png)

And in `pathlib`:

![image](https://user-images.githubusercontent.com/960340/166457656-6a9067ca-2338-4468-aee0-a4c07f1289d8.png)